packages/@aws-cdk/core/lib/private/synthesis.ts

```diff
@@ -49,7 +57,103 @@ export function synthesize(root: IConstruct, options: SynthesisOptions = { }): c
   // stacks to add themselves to the synthesized cloud assembly.
   synthesizeTree(root, builder, options.validateOnSynthesis);
 
-  return builder.buildAssembly();
+  const assembly = builder.buildAssembly();
+
+  invokeValidationPlugins(root, builder.outdir, assembly);
+
+  return assembly;
+}
+
+class LazyHash {
+  private hash?: string;
+
+  constructor(private readonly outdir: string) {}
+
+  get value(): string {
+    if (!this.hash) {
+      this.hash = computeChecksumOfFolder(this.outdir);
+    }
+    return this.hash;
+  }
+}
+
+function invokeValidationPlugins(root: IConstruct, outdir: string, assembly: CloudAssembly) {
+  const lazyHash = new LazyHash(outdir);
+
+  const templatePathsByPlugin: Map<IPolicyValidationPlugin, string[]> = new Map();
+  visit(root, 'post', construct => {
+    if (Stage.isStage(construct)) {
+      for (const plugin of construct.validationPlugins) {
+        if (!templatePathsByPlugin.has(plugin)) {
+          templatePathsByPlugin.set(plugin, []);
+        }
+        templatePathsByPlugin.get(plugin)!.push(...assembly.stacks.map(stack => stack.templateFullPath));
+      }
+    }
+  });
+
+  const reports: NamedValidationPluginReport[] = [];
+  for (const [plugin, paths] of templatePathsByPlugin.entries()) {
+    const originalHash = lazyHash.value;
+
+    try {
+      const report = plugin.validate({ templatePaths: paths });
+      reports.push({ ...report, pluginName: plugin.name });
+    } catch (e) {
+      reports.push({
+        success: false,
+        pluginName: plugin.name,
+        violations: [],
+        metadata: {
+          error: `Validation plugin '${plugin.name}' failed: ${e.message}`,
+        },
+      });
+    }
+    if (computeChecksumOfFolder(outdir) !== originalHash) {
+      throw new Error(`Illegal operation: validation plugin '${plugin.name}' modified the cloud assembly`);
+    }
+  }
+
+  const failed = reports.some(r => !r.success);
+  if (failed) {
+    const tree = new ConstructTree(root);
+    const formatter = new PolicyValidationReportFormatter(tree);
+    const formatJson = FeatureFlags.of(root).isEnabled(cxapi.VALIDATION_REPORT_JSON) ?? false;
+    const output = formatJson
+      ? formatter.formatJson(reports)
+      : formatter.formatPrettyPrinted(reports);
+
+    if (formatJson) {
+      // eslint-disable-next-line no-console
+      console.log(JSON.stringify(output, undefined, 2));
+    } else {
+      // eslint-disable-next-line no-console
+      console.error(output);
+    }
+
+    throw new Error('Validation failed. See the validation report above for details');
+  }
```

Comment:
Do you think we should just print the overall status like

```
Validation successful!
```

Reply:
Ok I updated the report so that it now prints out this if there are failures

![Screenshot 2023-03-22 at 11 19 49 AM](https://user-images.githubusercontent.com/43035978/226952190-2f04e7de-bbd9-4b78-b79d-a1382870fdb9.png)


And this if there are no failures.

![Screenshot 2023-03-22 at 11 21 25 AM](https://user-images.githubusercontent.com/43035978/226952677-8e82e75a-cb0a-4629-a1fb-1813d5781c13.png)

Reply:
We now always print the summary at the end with _all_ the plugins that were executed.